File: TBLib/Utilities.swift
import System
import System.Collections.Generic
import System.Linq
import System.Text
import System.Threading.Tasks
import UnityEngine
import UnityModManagerNet
import UnityEngine.UI
import HarmonyLib
import Kingmaker
import Kingmaker.AI.Blueprints
import Kingmaker.AI.Blueprints.Considerations
import Kingmaker.AreaLogic.Cutscenes
import Kingmaker.AreaLogic.Etudes
import Kingmaker.Armies
import Kingmaker.Armies.TacticalCombat
import Kingmaker.Armies.TacticalCombat.Blueprints
import Kingmaker.Armies.TacticalCombat.Brain
import Kingmaker.Armies.TacticalCombat.Brain.Considerations
import Kingmaker.BarkBanters
import Kingmaker.Blueprints
import Kingmaker.Blueprints.Area
import Kingmaker.Blueprints.CharGen
import Kingmaker.Blueprints.Classes
import Kingmaker.Blueprints.Credits
import Kingmaker.Blueprints.Encyclopedia
import Kingmaker.Blueprints.Facts
import Kingmaker.Blueprints.Items
import Kingmaker.Blueprints.Items.Ecnchantments
import Kingmaker.Blueprints.Items.Armors
import Kingmaker.Blueprints.Items.Components
import Kingmaker.Blueprints.Items.Equipment
import Kingmaker.Blueprints.Items.Shields
import Kingmaker.Blueprints.Items.Weapons
import Kingmaker.Kingdom.Blueprints
import Kingmaker.Kingdom.Settlements
import Kingmaker.Blueprints.Quests
import Kingmaker.Blueprints.Root
import Kingmaker.Cheats
import Kingmaker.Blueprints.Console
import Kingmaker.Controllers.Rest
import Kingmaker.Designers
import Kingmaker.DialogSystem.Blueprints
import Kingmaker.Dungeon.Blueprints
import Kingmaker.EntitySystem.Entities
import Kingmaker.EntitySystem.Stats
import Kingmaker.GameModes
import Kingmaker.Globalmap.Blueprints
import Kingmaker.Interaction
import Kingmaker.Items
import Kingmaker.PubSubSystem
import Kingmaker.RuleSystem
import Kingmaker.RuleSystem.Rules.Damage
import Kingmaker.Tutorial
import Kingmaker.UI
import Kingmaker.UI.Common
import Kingmaker.UnitLogic
import Kingmaker.UnitLogic.Buffs
import Kingmaker.UnitLogic.Buffs.Blueprints
import Kingmaker.UnitLogic.Customization
import Kingmaker.Utility
import Kingmaker.Visual.Sound

public class NamedTypeFilter {
    public var name: String
    public var type: Type

    public init(_ name: String, type: Type) {
        self.name = name
        self.type = type
    }
}

File: ToyBox/classes/Infrastructure/Utilities.cs
// Copyright < 2021 > Narria(github user Cabarius) - License: MIT
using UnityEngine;
using UnityModManagerNet;
using UnityEngine.UI;
using UnityEditor;
using HarmonyLib;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using System.Text.RegularExpressions;
using System.Threading.Tasks;
using System.Xml.Serialization;
using Kingmaker;
using Kingmaker.AI.Blueprints;
using Kingmaker.AI.Blueprints.Considerations;
using Kingmaker.AreaLogic.Cutscenes;
using Kingmaker.AreaLogic.Etudes;
using Kingmaker.Armies;
using Kingmaker.Armies.TacticalCombat;
using Kingmaker.Armies.TacticalCombat.Blueprints;
using Kingmaker.Armies.TacticalCombat.Brain;
using Kingmaker.Armies.TacticalCombat.Brain.Considerations;
using Kingmaker.BarkBanters;
using Kingmaker.Blueprints;
using Kingmaker.Blueprints.Area;
using Kingmaker.Blueprints.CharGen;
using Kingmaker.Blueprints.Classes;
using Kingmaker.Blueprints.Credits;
using Kingmaker.Blueprints.Encyclopedia;
using Kingmaker.Blueprints.Facts;
using Kingmaker.Blueprints.Items;
using Kingmaker.Blueprints.Items.Armors;
using Kingmaker.Blueprints.Items.Components;
using Kingmaker.Blueprints.Items.Equipment;
using Kingmaker.Blueprints.Items.Shields;
using Kingmaker.Blueprints.Items.Weapons;
using Kingmaker.Blueprints.Quests;
using Kingmaker.Blueprints.Root;
using Kingmaker.Cheats;
using Kingmaker.Blueprints.Console;
using Kingmaker.Controllers.Rest;
using Kingmaker.Designers;
using Kingmaker.DialogSystem.Blueprints;
using Kingmaker.Dungeon.Blueprints;
using Kingmaker.EntitySystem.Entities;
using Kingmaker.EntitySystem.Stats;
using Kingmaker.GameModes;
using Kingmaker.Globalmap.Blueprints;
using Kingmaker.Interaction;
using Kingmaker.Items;
using Kingmaker.PubSubSystem;
using Kingmaker.RuleSystem;
using Kingmaker.RuleSystem.Rules.Damage;
using Kingmaker.Tutorial;
using Kingmaker.UI;
using Kingmaker.UI.Common;
using Kingmaker.UnitLogic;
using Kingmaker.UnitLogic.Buffs;
using Kingmaker.UnitLogic.Buffs.Blueprints;
using Kingmaker.UnitLogic.Customization;
using Kingmaker.UnitLogic.Mechanics;
using Kingmaker.Utility;
using Kingmaker.Visual.Sound;
using Kingmaker.Assets.UI;

namespace ToyBox {
    public static class Utilties {

        public static TValue GetValueOrDefault<TKey, TValue>(this Dictionary<TKey, TValue> dictionary, TKey key, TValue defaultValue = default(TValue)) {
            if (dictionary == null) { throw new ArgumentNullException(nameof(dictionary)); } // using C# 6
            if (key == null) { throw new ArgumentNullException(nameof(key)); } //  using C# 6

            TValue value;
            return dictionary.TryGetValue(key, out value) ? value : defaultValue;
        }
        public static object GetPropValue(this object obj, String name) {
            foreach (String part in name.Split('.')) {
                if (obj == null) { return null; }

                Type type = obj.GetType();
                PropertyInfo info = type.GetProperty(part);
                if (info == null) { return null; }

                obj = info.GetValue(obj, null);
            }
            return obj;
        }
        public static T GetPropValue<T>(this object obj, String name) {
            object retval = GetPropValue(obj, name);
            if (retval == null) { return default(T); }
            // throws InvalidCastException if types are incompatible
            return (T)retval;
        }
        public static object SetPropValue(this object obj, String name, object value) {
            var parts = name.Split('.');
            var final = parts.Last();
            if (final == null) return null;
            foreach (String part in parts) {
                if (obj == null) { return null; }
                Type type = obj.GetType();
                PropertyInfo info = type.GetProperty(part);
                if (info == null) { return null; }
                if (part == final) {
                    info.SetValue(obj, value);
                    return value;
                }
                else {
                    obj = info.GetValue(obj, null);
                }
            }
            return null;
        }
        public static T SetPropValue<T>(this object obj, String name, T value) {
            object retval = SetPropValue(obj, name, value);
            if (retval == null) { return default(T); }
            // throws InvalidCastException if types are incompatible
            return (T)retval;
        }
        public static string RemoveHtmlTags(string s) {
            return Regex.Replace(s, "<.*?>", String.Empty);
        }
        public static string UnityRichTextToHtml(string s) {
            s = s.Replace("<color=", "<font color=");
            s = s.Replace("</color>", "</font>");
            s = s.Replace("<size=", "<size size=");
            s = s.Replace("</size>", "</font>");
            s += "<br/>";

            return s;
        }

        public static string[] getObjectInfo(object o) {

            string fields = "";
            foreach (string field in Traverse.Create(o).Fields()) {
                fields = fields + field + ", ";
            }
            string methods = "";
            foreach (string method in Traverse.Create(o).Methods()) {
                methods = methods + method + ", ";
            }
            string properties = "";
            foreach (string property in Traverse.Create(o).Properties()) {
                properties = properties + property + ", ";
            }
            return new string[] { fields, methods, properties };
        }
    }
    public static class TB {
        public static bool IsKindOf(this Type type, Type baseType) {
            return type.IsSubclassOf(baseType) || type == baseType;
        }
    }

    public static class BlueprintScriptableObjectUtils {
        public static string GetDescription(this BlueprintScriptableObject bpObejct)
        // borrowed shamelessly and enchanced from Bag of Tricks https://www.nexusmods.com/pathfinderkingmaker/mods/26, which is under the MIT License
        {
            try {
                UnitReference mainChar = Game.Instance.Player.MainCharacter;
                if (mainChar == null) { return ""; }
                MechanicsContext context = new MechanicsContext((UnitEntityData)null, mainChar.Value.Descriptor, bpObejct, (MechanicsContext)null, (TargetWrapper)null);
                return context?.SelectUIData(UIDataType.Description)?.Description ?? "";
            }
            catch (Exception e) {
                Console.Write($"{e}");
#if DEBUG
                return "ERROR".red().bold() + $": caught exception {e}";
#else
                return "";
#endif
            }
        }
    }
}
